### DISCUSSION TITLE
Documentation improvement for SQL beginners

### DISCUSSION THREAD

**Discussion by Alex-V69** [sqlalchemy/sqlalchemy#11932] [Usage Questions] [answered]
I was reading [SQLAlchemy Unified Tutorial](https://docs.sqlalchemy.org/en/20/tutorial/index.html) until I was offered to learn more about [Session.](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#basics-of-using-a-session)
When I reached the [deleting](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#deleting) section, I realized what a rabbit hole I ventured into because no mater how many times I read this paragraph:

> _"Rows that correspond to mapped objects that are related to a deleted object via the [relationship()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship) directive are not deleted by default. If those objects have a foreign key constraint back to the row being deleted, those columns are set to NULL. This will cause a constraint violation if the columns are non-nullable."_

I couldn't comprehend it.

P.S. I have no prior experience working with SQL.

**Comment by gordthompson** [comment, marked as answer]
As you become more familiar with SQL, specifically *foreign keys* and (ORM) *relationships*, that passage will make more sense. It describes the default behaviour when a deleted object has a relationship with another object class.

Consider a company that manages vacation condos. They have `Rental` condos and a `Renter` can rent one or more of them.

```python
class Renter(Base):
    __tablename__ = "renter"
    renter_id = Column(String(20), primary_key=True)
    renter_name = Column(String(50), nullable=False)
    rentals = relationship("Rental")


class Rental(Base):
    __tablename__ = "rental"
    rental_id = Column(String(20), primary_key=True)
    rental_name = Column(String(50), nullable=False)
    renter_id = Column(String(20), ForeignKey("renter.renter_id"))
```

Notice that `Renter` includes `rentals = relationship("Rental")` so we can retrieve a list of the `Rental` objects that a given customer is renting.

If in the database we have

```
[renter]

renter_id  renter_name
---------  -----------
gt01       Gord T     
mb01       Mike B     


[rental]

rental_id  rental_name     renter_id
---------  --------------  ---------
miami01    Miami condo #1  gt01     
```

and we delete me as a `Renter`

```python
with Session(engine) as sess:
    gord = sess.get(Renter, "gt01")
    sess.delete(gord)
    sess.commit()
```

then by default the 'miami01' `Rental` is not deleted; it has its `renter_id` set to NULL.

```
[renter]

renter_id  renter_name
---------  -----------
mb01       Mike B     


[rental]

rental_id  rental_name     renter_id
---------  --------------  ---------
miami01    Miami condo #1           

```

**Reply by CaselIT** [reply]
To control this behaviour relationship has a cascade parameter. It's documented here https://docs.sqlalchemy.org/en/20/orm/cascades.html#unitofwork-cascades

**Comment by CaselIT** [comment]
@Alex-V69 do you have an alternative wording that you think could have made it easier to understand?

**Reply by Alex-V69** [reply]
I don't because I don't have a prior experience working with SQL. I hoped that I wouldn't have to know it, since it's hidden behind high level abstractions of SQL Alchemy.

**Reply by zzzeek** [reply]
So this is unfortunately not SQLAlchemy's philosophy.   SQLAlchemy has zero intention of "hiding" SQL, instead it seeks to automate the production and interaction with SQL.   imagine if you went to a factory that puts soda into bottles.   You'd see thousands of soda bottles being filled.  Is the activity here being "hidden" ?  Do the people who operate the soda bottling machine themselves have no knowledge of how to put soda into a bottle ?  Of course not.  The machine doesnt intend to "hide" any of this, that would not even be useful.  It just *automates* it for scale.   

If you are looking for tools to hide the details of SQL I would look at Django ORM and PeeWee ORM.

**Comment by Alex-V69** [comment]
"PeeWee ORM" – That looks interesting! Thank you for your suggestion, but I'm not going to give up just yet. 😄
I just wanted to suggest how the docs can be improved. For example, there could be a new section for a total beginners like me, or every article could have a "What you need to know to follow" section with useful links where possible.
